swarm/src/handler/pending.rs

```diff
@@ -0,0 +1,120 @@
+// Copyright 2022 Protocol Labs.
+// Copyright 2018 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use crate::handler::{
+    ConnectionHandler, ConnectionHandlerEvent, ConnectionHandlerUpgrErr, KeepAlive,
+    SubstreamProtocol,
+};
+use crate::NegotiatedSubstream;
+use libp2p_core::{
+    upgrade::{InboundUpgrade, OutboundUpgrade, PendingUpgrade},
+    Multiaddr,
+};
+use std::task::{Context, Poll};
+use void::Void;
+
+/// Implementation of [`ConnectionHandler`] that returns a pending upgrade.
+#[derive(Clone, Debug)]
+pub struct PendingConnectionHandler {
+    protocol_name: String,
+}
+
+impl PendingConnectionHandler {
+    pub fn new(protocol_name: String) -> Self {
+        PendingConnectionHandler { protocol_name }
+    }
+}
+
+impl ConnectionHandler for PendingConnectionHandler {
+    type InEvent = Void;
+    type OutEvent = Void;
+    type Error = Void;
+    type InboundProtocol = PendingUpgrade<String>;
+    type OutboundProtocol = PendingUpgrade<String>;
+    type OutboundOpenInfo = Void;
+    type InboundOpenInfo = ();
+
+    fn listen_protocol(&self) -> SubstreamProtocol<Self::InboundProtocol, Self::InboundOpenInfo> {
+        SubstreamProtocol::new(PendingUpgrade::new(self.protocol_name.clone()), ())
+    }
+
+    fn inject_fully_negotiated_inbound(
+        &mut self,
+        protocol: <Self::InboundProtocol as InboundUpgrade<NegotiatedSubstream>>::Output,
+        _: Self::InboundOpenInfo,
+    ) {
+        void::unreachable(protocol)
+    }
+
+    fn inject_fully_negotiated_outbound(
+        &mut self,
+        protocol: <Self::OutboundProtocol as OutboundUpgrade<NegotiatedSubstream>>::Output,
+        _info: Self::OutboundOpenInfo,
+    ) {
+        void::unreachable(protocol);
+        #[allow(unreachable_code)]
+        {
+            void::unreachable(_info);
+        }
```

Comment:
```suggestion
        _: Self::OutboundOpenInfo,
    ) {
        void::unreachable(protocol);
```

I'd almost suggest to not use `void::unreachable` for the `info` part if we have to `#[allow(unreachable_code)]` for it to work.

Reply:
I prefer the additional `void::unreachable`, just in case we refactor this part of the codebase in the future.